libmamba/src/core/url.cpp

```diff
@@ -15,6 +17,152 @@
 
 namespace mamba
 {
+
+    namespace
+    {
+        /**
+         * A RAII `CURL*` createed from `curl_easy_handle`.
+         *
+         * Never null, throw exception at construction if creating the handle fails.
+         */
+        class CurlEasyHandle
+        {
+        public:
+
+            using value_type = ::CURL;
+            using pointer = value_type*;
+            using const_pointer = const value_type*;
+
+            CurlEasyHandle();
+            ~CurlEasyHandle();
+
+            [[nodiscard]] auto raw() const -> const_pointer;
+            [[nodiscard]] auto raw() -> pointer;
+
+        private:
+
+            pointer m_handle = nullptr;
+        };
+
+        /**
+         * A RAII wrapper for string mananged by Curl.
+         *
+         * String can possibly be null, or zero-lenght, depending on the data returned by Curl.
+         */
+        class CurlStr
+        {
+            using value_type = char;
+            using pointer = value_type*;
+            using const_pointer = const value_type*;
+            using input_pointer = value_type**;
+            using size_type = int;
+
+        public:
+
+            explicit CurlStr() = default;
+            explicit CurlStr(pointer data);
+            explicit CurlStr(pointer data, size_type len);
+            ~CurlStr();
+
+            [[nodiscard]] auto raw_input() -> input_pointer;
+            [[nodiscard]] auto raw() const -> const_pointer;
+            [[nodiscard]] auto raw() -> pointer;
+
+            [[nodiscard]] auto str() const -> std::optional<std::string_view>;
+
+        private:
+
+            pointer m_data = nullptr;
+            // Only meaningful when > 0, otherwise, assume null terminated string
+            size_type m_len = -1;
+
+            void free_data_if_present();
+        };
+
+        CurlEasyHandle::CurlEasyHandle()
+        {
+            m_handle = ::curl_easy_init();
+            if (m_handle == nullptr)
+            {
+                throw std::runtime_error("Could not create CURL handle");
+            }
+        }
+
+        CurlEasyHandle::~CurlEasyHandle()
+        {
+            ::curl_easy_cleanup(m_handle);
+        }
+
+        auto CurlEasyHandle::raw() const -> const_pointer
+        {
+            return m_handle;
+        }
+
+        auto CurlEasyHandle::raw() -> pointer
+        {
+            return m_handle;
+        }
+
+        CurlStr::CurlStr(pointer data)
+            : m_data(data)
+        {
+        }
+
+        CurlStr::CurlStr(pointer data, size_type len)
+            : m_data(data)
+            , m_len(len)
+        {
+        }
+
+        CurlStr::~CurlStr()
+        {
+            free_data_if_present();
+        }
+
+        auto CurlStr::raw_input() -> input_pointer
+        {
+            free_data_if_present();
```

Comment:
I think this should not be called here

Reply:
Should we throw instead if there is data present?

Reply:
If it's a development mistake when the data is not present, put an assert.